### PR TITLE
Add API for testing netgroups

### PR DESCRIPTION
**Pull request by pbrezina**
This PR also contains to fixes for the get method.

### CODE REVIEW

**Comment by justin-stephenson** [comment]
Using the AnyProvider topology group fails with IPA because the domain name is also returned in the `getent netgroup ng-1` output (by the way I wonder if it makes sense to add the example AnyProvider test into sssd repo alongside these new API addition PRs, it was simple but I had to write it myself based on the code comment example to test this).

~~~
(.venv) [justin@fedora system]$ cat tests/test_netgroup.py 
from __future__ import annotations

import re
import time
from datetime import datetime, timedelta

import pytest
from sssd_test_framework.roles.ad import AD
from sssd_test_framework.roles.client import Client
from sssd_test_framework.roles.generic import GenericADProvider, GenericProvider
from sssd_test_framework.roles.ldap import LDAP
from sssd_test_framework.roles.samba import Samba
from sssd_test_framework.topology import KnownTopology, KnownTopologyGroup

@pytest.mark.topology(KnownTopologyGroup.AnyProvider)
def test_example_netgroup(client: Client, provider: GenericProvider):
        # Create user
        user = provider.user("user-1").add()
        # Create two netgroups
        ng1 = provider.netgroup("ng-1").add()
        ng2 = provider.netgroup("ng-2").add()
        # Add user and ng2 as members to ng1
        ng1.add_member(user=user)
        ng1.add_member(ng=ng2)
        # Add host as member to ng2
        ng2.add_member(host="client")
        # Start SSSD
        client.sssd.start()
        # Call `getent netgroup ng-1` and assert the results
        result = client.tools.getent.netgroup("ng-1")
        print(result.members)
        assert result is not None
        assert result.name == "ng-1"
        assert len(result.members) == 2
        assert "(-,user-1,)" in result.members
        assert "(client,-,)" in result.members


(.venv) [justin@fedora system]$ pytest --mh-config=mhc.yaml --mh-log-path=mh.log -v tests/test_netgroup.py
FAILED tests/test_netgroup.py::test_example_netgroup (ipa) - AssertionError: assert '(-,user-1,)' in ['(-,user-1,ipa.test)', '(client.test,-,ipa.test)']                                                                                      
======================================================================================================== 1 failed, 3 passed in 50.15s ========================================================================================================

File: sssd_test_framework/roles/ad.py [hunk 1]
```diff
@@ -382,6 +384,8 @@ def _exec(self, op: str, args: list[str] | str | None = None, **kwargs) -> SSHPr
         :type op: str
         :param args: List of additional command arguments, defaults to None
         :type args: list[str] | None, optional
+        :param format_with: Command that will be used to format the output (e.g. Format-List), defaults to None
```
Comment (ikerexxe) [inline sssd_test_framework/roles/ad.py:431]: Can you explain what `None` means for this case?

**Comment by pbrezina** [comment]
> Using the AnyProvider topology group fails with IPA because the domain name is also returned in the `getent netgroup ng-1` output (by the way I wonder if it makes sense to add the example AnyProvider test into sssd repo alongside these new API addition PRs, it was simple but I had to write it myself based on the code comment example to test this).

Unfortunately, IPA behaves differently. The administrator does not really have control on how the netgroup triple looks like, there are three things:

- IPA automatically adds domain
- IPA converts existing hosts to the fully qualified name
- IPA generate triples for all host-user combination

```
$ ipa netgroup-add ng-1
$ ipa netgroup-add-member ng1 --users admin
$ ipa netgroup-add-member ng1 --hosts client
$ ipa netgroup-add-member ng1 --hosts test
$ getent netgroup ng-1
ng-1                  (client.test,admin,ipa.test) (test,admin,ipa.test)
```

It is possible to use AnyProvider, but you have to know the difference and add the netgroup so it follows IPA rules. Do you have any API suggestion on how to solve this neatly? Otherwise, I plan only to document it on how to guide.

**Review by ikerexxe — DISMISSED** [review]
LGTM!

**Comment by pbrezina** [comment]
> > It is possible to use AnyProvider, but you have to know the difference and add the netgroup so it follows IPA rules. Do you have any API suggestion on how to solve this neatly? Otherwise, I plan only to document it on how to guide.
> 
> Thanks for clarification, IMO we just need to document this clearly.

I had an epiphany and figured out how to nicely ignore the domain part if the test allows it. For example, this is test for [this bugzilla](https://bugzilla.redhat.com/show_bug.cgi?id=2162552), where it is safe to ignore the domain part. And now we know with just single test that the problem happens only with IPA provider, other providers are green.

```python
@pytest.mark.topology(KnownTopologyGroup.AnyProvider)
def test_example_netgroup(client: Client, provider: GenericProvider):
    u1 = provider.user("user-1").add()
    u2 = provider.user("user-2").add()

    ng1 = provider.netgroup("ng-1").add().add_member(user=u1)
    ng2 = provider.netgroup("ng-2").add().add_member(user=u2, ng=ng1)

    client.sssd.start()

    result = client.tools.getent.netgroup("ng-2")
    assert result is not None
    assert result.name == "ng-2"
    assert len(result.members) == 2
    assert "(-, user-1)" in result.members
    assert "(-, user-2)" in result.members

    ng2.remove_member(ng=ng1)
    client.host.ssh.run("sss_cache -E")

    result = client.tools.getent.netgroup("ng-2")
    assert result is not None
    assert result.name == "ng-2"
    assert len(result.members) == 1
    assert "(-, user-1)" not in result.members
    assert "(-, user-2)" in result.members
```

I also extended documentation for NetgroupEntry and added "How to test netgroups" guide.

**Review by justin-stephenson — APPROVED** [review]

**Comment by justin-stephenson** [comment]
> > > It is possible to use AnyProvider, but you have to know the difference and add the netgroup so it follows IPA rules. Do you have any API suggestion on how to solve this neatly? Otherwise, I plan only to document it on how to guide.
> > 
> > 
> > Thanks for clarification, IMO we just need to document this clearly.
> 
> I had an epiphany and figured out how to nicely ignore the domain part if the test allows it. For example, this is test for [this bugzilla](https://bugzilla.redhat.com/show_bug.cgi?id=2162552), where it is safe to ignore the domain part. And now we know with just single test that the problem happens only with IPA provider, other providers are green.
> 
> ```python
> @pytest.mark.topology(KnownTopologyGroup.AnyProvider)
> def test_example_netgroup(client: Client, provider: GenericProvider):
>     u1 = provider.user("user-1").add()
>     u2 = provider.user("user-2").add()
> 
>     ng1 = provider.netgroup("ng-1").add().add_member(user=u1)
>     ng2 = provider.netgroup("ng-2").add().add_member(user=u2, ng=ng1)
> 
>     client.sssd.start()
> 
>     result = client.tools.getent.netgroup("ng-2")
>     assert result is not None
>     assert result.name == "ng-2"
>     assert len(result.members) == 2
>     assert "(-, user-1)" in result.members
>     assert "(-, user-2)" in result.members
> 
>     ng2.remove_member(ng=ng1)
>     client.host.ssh.run("sss_cache -E")
> 
>     result = client.tools.getent.netgroup("ng-2")
>     assert result is not None
>     assert result.name == "ng-2"
>     assert len(result.members) == 1
>     assert "(-, user-1)" not in result.members
>     assert "(-, user-2)" in result.members
> ```
> 
> I also extended documentation for NetgroupEntry and added "How to test netgroups" guide.

Thanks, the docs look quite good. I commented on one minor typo, but I approve this PR.

File: docs/guides/testing-netgroups.rst [hunk 2]
```diff
@@ -0,0 +1,147 @@
...
+    assert ("client", "-") in result.members
+
+You probably want to use plain string in most scenarios as it is more readable
+and easier to write. But it may be nicer to use tuples if you use variables for
+the values instead of hard coded string.
+
+Topology parametrization with netgroups
+=======================================
+
+Active Directory, LDAP and Samba providers behaves exactly the same, however,
```
Comment (ikerexxe) [inline docs/guides/testing-netgroups.rst:92]: s/behaves/behave

**Review by ikerexxe — COMMENTED** [review]
I also have another minor typo, apart from that it looks good to me.

**Comment by pbrezina** [comment]
Fixed.

**Review by ikerexxe — APPROVED** [review]

**Pull request merged**